<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="UTF-8">
    <title>Правильный запуск нового проекта на Rust с цепочкой ошибок | Dudochkin Victor </title>
    <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
    <meta name="google-site-verification" content="Kepqb6k6uptlwngnHbBVJmqQtwNW8zfHyQBbTXLYAgA" />
    <meta name="yandex-verification" content="8e8bc73c2566b473" />
    <style>
    :root {
        /* Primary theme color */
        /* --primary-color: #121A26; */
        --primary-color: #0f1419;
        /* Primary theme text color */
        /* --primary-text-color: #121A26; */
        --primary-text-color: #0f1419;
        /* Primary theme link color */
        --primary-link-color: #053961;
        /* Secondary color: the background body color */
        --secondary-color: #F0F0F0;
        --secondary-text-color: #1C2834;
        /* Highlight text color of table of content */
        --toc-highlight-text-color: #053961;
    }
</style>
    
    
    <!-- Global site tag (gtag.js) - Google Analytics -->
    <script async src="https://www.googletagmanager.com/gtag/js?id=G-47MMPLLY44"></script>
    <script>
        window.dataLayer = window.dataLayer || [];
        function gtag(){dataLayer.push(arguments);}
        gtag('js', new Date());
        gtag('config', 'G-47MMPLLY44');
    </script>
    

    
    <!-- Yandex.Metrika counter -->
    <script type="text/javascript" >
        (function(m,e,t,r,i,k,a){m[i]=m[i]||function(){(m[i].a=m[i].a||[]).push(arguments)};
        m[i].l=1*new Date();k=e.createElement(t),a=e.getElementsByTagName(t)[0],k.async=1,k.src=r,a.parentNode.insertBefore(k,a)})
        (window, document, "script", "https://mc.yandex.ru/metrika/tag.js", "ym");
    
        ym('87064264', "init", {
            clickmap:true,
            trackLinks:true,
            accurateTrackBounce:true
        });
    </script>
    <noscript><div><img src="https://mc.yandex.ru/watch/87064264" style="position:absolute; left:-9999px;" alt="" /></div></noscript>
    <!-- /Yandex.Metrika counter -->
    
    
    <link rel="preconnect" href="https://fonts.gstatic.com">
    <link href="https://fonts.googleapis.com/css2?family=Roboto:wght@100;300;400&display=swap" rel="stylesheet">
    <link href="https://fonts.googleapis.com/icon?family=Material+Icons" rel="stylesheet">
    <link rel="stylesheet" href="/mdl.min.css">
    <link rel="stylesheet" href="https://dudochkin-victor.github.io/style.css">
    
    
</head>

<body>
    
<header class="box-shadow">
    

<a href="https:&#x2F;&#x2F;dudochkin-victor.github.io&#x2F;">
    <div class="logo">
        <img src="https://dudochkin-victor.github.io/ferris.png" alt="logo">
        Dudochkin Victor
    </div>
</a>

<nav>
    <!-- 
    <a class="nav-item subtitle-text" href="https:&#x2F;&#x2F;dudochkin-victor.github.io&#x2F;showcases&#x2F;">Showcases</a>
    
    <a class="nav-item subtitle-text" href="https:&#x2F;&#x2F;dudochkin-victor.github.io&#x2F;changelog&#x2F;">Changelog</a>
    
    <a class="nav-item subtitle-text" href="https:&#x2F;&#x2F;dudochkin-victor.github.io&#x2F;about&#x2F;">About</a>
     -->

    
        
        <a class="nav-item subtitle-text" href="&#x2F;solana&#x2F;">Solana</a>
        
        <a class="nav-item subtitle-text" href="&#x2F;edgeware&#x2F;">Edgeware</a>
        
        <a class="nav-item subtitle-text" href="&#x2F;blog&#x2F;">Блог</a>
        
        <a class="nav-item subtitle-text" href="https:&#x2F;&#x2F;github.com&#x2F;dudochkin-victor">Github</a>
        
    
</nav>

</header>

 
    <main>
        
        
        
        
        
        <div class="toc">
            <div class="toc-sticky">
                
                <div class="toc-item">
                    <a class="subtext" href="https://dudochkin-victor.github.io/blog/starting-a-new-rust-project-right-with-errorchain/#effektivnaia-tsepochka-oshibok">Эффективная цепочка ошибок</a>
                </div>
                
                
                <div class="toc-item">
                    <a class="subtext" href="https://dudochkin-victor.github.io/blog/starting-a-new-rust-project-right-with-errorchain/#tl-dr">TL;DR</a>
                </div>
                
                
            </div>
        </div>
        
        

        <div class="content text">
            
<h1>Правильный запуск нового проекта на Rust с цепочкой ошибок</h1>

<p><a href="https://brson.github.io/2016/11/30/starting-with-error-chain">Перевод</a> | Автор оригинала: Brian Anderson</p>
<p>Пока я готовил последний выпуск error-chain, меня осенило, что не все знают магию error-chain, маленькой библиотеки Rust, которая упрощает правильную обработку ошибок. Обработка ошибок является в Rust первоочередной задачей, она тонко сложна и важна для работоспособности проекта и здравомыслия его сопровождающих, чтобы обработка ошибок выполнялась правильно. С самого начала.</p>
<p>Первое, что я делаю - действительно, самое первое - при запуске любого проекта на Rust - это выбираю стратегию обработки ошибок. А с цепочкой ошибок я заранее точно знаю, что буду делать, и что это будет работать достаточно хорошо. Это дает мне душевное спокойствие.</p>
<p>Итак, этот пост призван продемонстрировать то, что я считаю основным передовым методом настройки обработки ошибок в Rust. Если вы с самого начала будете следовать этим простым инструкциям, у вас будут настроены все ваши леса ошибок, которые будут масштабироваться по мере роста вашего проекта во многих крэйтах, не требуя никаких дополнительных усилий.</p>
<p>Файл quickstart.rs цепочки ошибок демонстрирует простое, но мощное приложение, настроенное с помощью цепочки ошибок. Его можно использовать в качестве шаблона - просто скопируйте этот файл на свой main.rs, и у вас будет приложение, настроенное для надежной обработки ошибок.</p>
<p>Я воспроизведу его здесь полностью: </p>
<pre data-lang="rust" style="background-color:#0f1419;color:#bfbab0;" class="language-rust "><code class="language-rust" data-lang="rust"><span style="font-style:italic;color:#5c6773;">// Simple and robust error handling with error-chain!
</span><span style="font-style:italic;color:#5c6773;">// Use this as a template for new projects.
</span><span>
</span><span style="font-style:italic;color:#5c6773;">// `error_chain!` can recurse deeply
</span><span style="color:#bfbab0cc;">#!</span><span>[</span><span style="color:#ffb454;">recursion_limit </span><span style="color:#f29668;">= </span><span style="color:#c2d94c;">&quot;1024&quot;</span><span>]
</span><span>
</span><span style="font-style:italic;color:#5c6773;">// Import the macro. Don&#39;t forget to add `error-chain` in your
</span><span style="font-style:italic;color:#5c6773;">// `Cargo.toml`!
</span><span style="color:#bfbab0cc;">#</span><span>[</span><span style="color:#ffb454;">macro_use</span><span>] </span><span style="color:#ff7733;">extern crate</span><span> error_chain</span><span style="color:#bfbab0cc;">;
</span><span>
</span><span style="font-style:italic;color:#5c6773;">// We&#39;ll put our errors in an `errors` module, and other modules in
</span><span style="font-style:italic;color:#5c6773;">// this crate will `use errors::*;` to get access to everything
</span><span style="font-style:italic;color:#5c6773;">// `error_chain!` creates.
</span><span style="color:#ff7733;">mod </span><span style="color:#59c2ff;">errors </span><span>{
</span><span>    </span><span style="font-style:italic;color:#5c6773;">// Create the Error, ErrorKind, ResultExt, and Result types
</span><span>    </span><span style="color:#f07178;">error_chain! </span><span>{ }
</span><span>}
</span><span>
</span><span style="color:#ff7733;">use </span><span>errors</span><span style="color:#f29668;">::*</span><span style="color:#bfbab0cc;">;
</span><span>
</span><span style="color:#ff7733;">fn </span><span style="color:#ffb454;">main</span><span>() {
</span><span>    </span><span style="color:#ff7733;">if let </span><span style="font-style:italic;color:#39bae6;">Err</span><span>(</span><span style="color:#ff7733;">ref</span><span> e) </span><span style="color:#f29668;">= </span><span style="color:#f07178;">run</span><span>() {
</span><span>        </span><span style="color:#f07178;">println!</span><span>(</span><span style="color:#c2d94c;">&quot;error: </span><span style="color:#f29718;">{}</span><span style="color:#c2d94c;">&quot;</span><span style="color:#bfbab0cc;">,</span><span> e)</span><span style="color:#bfbab0cc;">;
</span><span>
</span><span>        </span><span style="color:#ff7733;">for</span><span> e </span><span style="color:#f29668;">in</span><span> e</span><span style="color:#f29668;">.</span><span style="color:#f07178;">iter</span><span>()</span><span style="color:#f29668;">.</span><span style="color:#f07178;">skip</span><span>(</span><span style="color:#f29718;">1</span><span>) {
</span><span>            </span><span style="color:#f07178;">println!</span><span>(</span><span style="color:#c2d94c;">&quot;caused by: </span><span style="color:#f29718;">{}</span><span style="color:#c2d94c;">&quot;</span><span style="color:#bfbab0cc;">,</span><span> e)</span><span style="color:#bfbab0cc;">;
</span><span>        }
</span><span>
</span><span>        </span><span style="font-style:italic;color:#5c6773;">// The backtrace is not always generated. Try to run this example
</span><span>        </span><span style="font-style:italic;color:#5c6773;">// with `RUST_BACKTRACE=1`.
</span><span>        </span><span style="color:#ff7733;">if let </span><span style="font-style:italic;color:#39bae6;">Some</span><span>(backtrace) </span><span style="color:#f29668;">=</span><span> e</span><span style="color:#f29668;">.</span><span style="color:#f07178;">backtrace</span><span>() {
</span><span>            </span><span style="color:#f07178;">println!</span><span>(</span><span style="color:#c2d94c;">&quot;backtrace: </span><span style="color:#f29718;">{:?}</span><span style="color:#c2d94c;">&quot;</span><span style="color:#bfbab0cc;">,</span><span> backtrace)</span><span style="color:#bfbab0cc;">;
</span><span>        }
</span><span>
</span><span>        </span><span style="color:#f29668;">::</span><span>std</span><span style="color:#f29668;">::</span><span>process</span><span style="color:#f29668;">::</span><span>exit(</span><span style="color:#f29718;">1</span><span>)</span><span style="color:#bfbab0cc;">;
</span><span>    }
</span><span>}
</span><span>
</span><span style="font-style:italic;color:#5c6773;">// Most functions will return the `Result` type, imported from the
</span><span style="font-style:italic;color:#5c6773;">// `errors` module. It is a typedef of the standard `Result` type
</span><span style="font-style:italic;color:#5c6773;">// for which the error type is always our own `Error`.
</span><span style="color:#ff7733;">fn </span><span style="color:#ffb454;">run</span><span>() </span><span style="color:#bfbab0cc;">-&gt; </span><span style="font-style:italic;color:#39bae6;">Result</span><span>&lt;()&gt; {
</span><span>    </span><span style="color:#ff7733;">use </span><span>std</span><span style="color:#f29668;">::</span><span>fs</span><span style="color:#f29668;">::</span><span>File</span><span style="color:#bfbab0cc;">;
</span><span>
</span><span>    </span><span style="font-style:italic;color:#5c6773;">// This operation will fail
</span><span>    File</span><span style="color:#f29668;">::</span><span>open(</span><span style="color:#c2d94c;">&quot;contacts&quot;</span><span>)
</span><span>        </span><span style="color:#f29668;">.</span><span style="color:#f07178;">chain_err</span><span>(|| </span><span style="color:#c2d94c;">&quot;unable to open contacts file&quot;</span><span>)</span><span style="color:#f29668;">?</span><span style="color:#bfbab0cc;">;
</span><span>
</span><span>    </span><span style="font-style:italic;color:#39bae6;">Ok</span><span>(())
</span><span>}
</span></code></pre>
<p>Это говорит само за себя, но я хочу отметить несколько моментов по этому поводу. Во-первых, эта основная функция: </p>
<pre data-lang="rust" style="background-color:#0f1419;color:#bfbab0;" class="language-rust "><code class="language-rust" data-lang="rust"><span style="color:#ff7733;">fn </span><span style="color:#ffb454;">main</span><span>() {
</span><span>    </span><span style="color:#ff7733;">if let </span><span style="font-style:italic;color:#39bae6;">Err</span><span>(</span><span style="color:#ff7733;">ref</span><span> e) </span><span style="color:#f29668;">= </span><span style="color:#f07178;">run</span><span>() {
</span><span>        </span><span style="color:#f07178;">println!</span><span>(</span><span style="color:#c2d94c;">&quot;error: </span><span style="color:#f29718;">{}</span><span style="color:#c2d94c;">&quot;</span><span style="color:#bfbab0cc;">,</span><span> e)</span><span style="color:#bfbab0cc;">;
</span><span>
</span><span>        </span><span style="color:#ff7733;">for</span><span> e </span><span style="color:#f29668;">in</span><span> e</span><span style="color:#f29668;">.</span><span style="color:#f07178;">iter</span><span>()</span><span style="color:#f29668;">.</span><span style="color:#f07178;">skip</span><span>(</span><span style="color:#f29718;">1</span><span>) {
</span><span>            </span><span style="color:#f07178;">println!</span><span>(</span><span style="color:#c2d94c;">&quot;caused by: </span><span style="color:#f29718;">{}</span><span style="color:#c2d94c;">&quot;</span><span style="color:#bfbab0cc;">,</span><span> e)</span><span style="color:#bfbab0cc;">;
</span><span>        }
</span><span>
</span><span>        </span><span style="font-style:italic;color:#5c6773;">// The backtrace is not always generated. Try to run this example
</span><span>        </span><span style="font-style:italic;color:#5c6773;">// with `RUST_BACKTRACE=1`.
</span><span>        </span><span style="color:#ff7733;">if let </span><span style="font-style:italic;color:#39bae6;">Some</span><span>(backtrace) </span><span style="color:#f29668;">=</span><span> e</span><span style="color:#f29668;">.</span><span style="color:#f07178;">backtrace</span><span>() {
</span><span>            </span><span style="color:#f07178;">println!</span><span>(</span><span style="color:#c2d94c;">&quot;backtrace: </span><span style="color:#f29718;">{:?}</span><span style="color:#c2d94c;">&quot;</span><span style="color:#bfbab0cc;">,</span><span> backtrace)</span><span style="color:#bfbab0cc;">;
</span><span>        }
</span><span>
</span><span>        </span><span style="color:#f29668;">::</span><span>std</span><span style="color:#f29668;">::</span><span>process</span><span style="color:#f29668;">::</span><span>exit(</span><span style="color:#f29718;">1</span><span>)</span><span style="color:#bfbab0cc;">;
</span><span>    }
</span><span>}
</span></code></pre>
<p>Это типично для основных функций, которые я пишу в последнее время. Вся цель состоит в том, чтобы немедленно делегировать функции функции, которая участвует в обработке ошибок (возвращает наши настраиваемые типы результатов и ошибок), а затем обрабатывать эти ошибки. Эта процедура обработки ошибок демонстрирует три части информации, которую цепочка ошибок передает из ошибки: ближайшая ошибка, здесь привязка e; причинная цепочка ошибок, которая привела к этой ошибке; и след исходной ошибки. В зависимости от вашего варианта использования вы можете не беспокоиться об обратных трассировках или можете добавить вызов catch_unwind для борьбы с паниками.</p>
<p>Если вы запустите этот пример, вы увидите следующий результат: </p>
<pre style="background-color:#0f1419;color:#bfbab0;"><code><span>error: unable to open contacts file
</span><span>caused by: The system cannot find the file specified. (os error 2)
</span></code></pre>
<p>Это демонстрирует смысл существования цепочки ошибок: фиксация цепочки множественных ошибок, приведших к сбою программы, и создание отчетов о ней. Причина, по которой мы видим как окончательную ошибку, так и ее причину, заключается в том, что в нашей функции выполнения мы связали две ошибки вместе: </p>
<pre data-lang="rust" style="background-color:#0f1419;color:#bfbab0;" class="language-rust "><code class="language-rust" data-lang="rust"><span style="color:#ff7733;">fn </span><span style="color:#ffb454;">run</span><span>() </span><span style="color:#bfbab0cc;">-&gt; </span><span style="font-style:italic;color:#39bae6;">Result</span><span>&lt;()&gt; {
</span><span>    </span><span style="color:#ff7733;">use </span><span>std</span><span style="color:#f29668;">::</span><span>fs</span><span style="color:#f29668;">::</span><span>File</span><span style="color:#bfbab0cc;">;
</span><span>
</span><span>    </span><span style="font-style:italic;color:#5c6773;">// This operation will fail
</span><span>    File</span><span style="color:#f29668;">::</span><span>open(</span><span style="color:#c2d94c;">&quot;contacts&quot;</span><span>)
</span><span>        </span><span style="color:#f29668;">.</span><span style="color:#f07178;">chain_err</span><span>(|| </span><span style="color:#c2d94c;">&quot;unable to open contacts file&quot;</span><span>)</span><span style="color:#f29668;">?</span><span style="color:#bfbab0cc;">;
</span><span>
</span><span>    </span><span style="font-style:italic;color:#39bae6;">Ok</span><span>(())
</span><span>}
</span></code></pre>
<p>Этот вызов chain_err вызвал новую ошибку с нашим сообщением об ошибке, специфичным для нашего приложения, при сохранении исходной ошибки, сгенерированной File::open, в цепочке ошибок. Это простой пример, но вы можете представить себе, что в более крупных приложениях цепочка ошибок может быть довольно подробной.</p>
<p>Вот как вы начинаете работу с цепочкой ошибок, но это еще не все. Подробнее читайте в документации.</p>
<h2 id="effektivnaia-tsepochka-oshibok">Эффективная цепочка ошибок</h2>
<p>Несколько замечаний о том, как я использую цепочку ошибок.</p>
<p>Когда я занимаюсь хакерством, я без колебаний использую строки как ошибки, которые могут быть легко сгенерированы ala bail! («Мне это не нравится: {}», чепуха). Они представлены как ErrorKind::Msg, вариант, определенный для всех типов ошибок, генерируемых цепочкой ошибок! макрос.</p>
<p>Для приложений строки часто подходят как тип ошибки. Однако, когда вы разрабатываете API для публичного использования, именно это становится важным при определении типов ошибок (с помощью блока error_chain! Errors {}). Введенные варианты ошибок дают потребителям вашей библиотеки что-то подходящее. error-chain дает вам возможность делать простые или сложные вещи, она масштабируется в соответствии с потребностями вашего кода.</p>
<p>Ставьте свою error_chain! вызов внутри модуля ошибок и импортировать все содержимое с помощью errors::*. Импорт глобусов - это не то, чем вы хотите много заниматься, но в этом случае закономерность того стоит: вы действительно хотите, чтобы эти четыре типа были под рукой в каждом модуле крэйта.</p>
<p>Я стараюсь не слишком полагаться на автоматическое преобразование из foreign_links {}. Внешние ссылки автоматически преобразуются в локальный тип ошибки. Их легко настроить, и они позволяют легко взаимодействовать с ошибками, не зависящими от вас, но, выполняя автоматическое преобразование, вы теряете возможность вернуть ошибку, более актуальную для вашего приложения. То есть, вместо того, чтобы возвращать ошибку «система не может найти указанный файл», я хочу вернуть ошибку «не удалось открыть файл контактов», вызванную тем, что «система не может найти указанный файл». Каждая ссылка в ошибке поясняет, что пошло не так. Так что вместо того, чтобы использовать? при внешней ошибке используйте chain_err, чтобы дать больше контекста.</p>
<p>error-chain действительно сияет, когда вы начинаете создавать созвездие крэйтов, использующих стратегию цепочки ошибок, и все они связаны друг с другом через error_chain! ссылки {} блоки. Ошибки связанных цепочек ошибок могут распространяться по обратным трассам и иметь структурную форму, которую легко сопоставить, так что, например, вашу ошибку, которая возникла в вашем крэйте утилит, пузырилась через ваш сетевой крэйт, а затем поднималась по крэйту вашего приложения, легко определить с помощью сопоставления с образцом, например: </p>
<pre data-lang="rust" style="background-color:#0f1419;color:#bfbab0;" class="language-rust "><code class="language-rust" data-lang="rust"><span style="font-style:italic;color:#5c6773;">// Imagine these are crates, not mods
</span><span style="color:#ff7733;">mod </span><span style="color:#59c2ff;">utils </span><span>{
</span><span>    </span><span style="color:#f07178;">error_chain! </span><span>{
</span><span>        errors { Parse }
</span><span>    }
</span><span>}
</span><span>
</span><span style="color:#ff7733;">mod </span><span style="color:#59c2ff;">net </span><span>{
</span><span>    </span><span style="color:#f07178;">error_chain! </span><span>{
</span><span>        links {
</span><span>            Utils(</span><span style="color:#f29668;">::</span><span>utils</span><span style="color:#f29668;">::</span><span>Error</span><span style="color:#bfbab0cc;">, </span><span style="color:#f29668;">::</span><span>utils</span><span style="color:#f29668;">::</span><span>ErrorKind)</span><span style="color:#bfbab0cc;">;
</span><span>        }
</span><span>    }
</span><span>}
</span><span>
</span><span style="color:#ff7733;">mod </span><span style="color:#59c2ff;">app </span><span>{
</span><span>    </span><span style="color:#f07178;">error_chain! </span><span>{
</span><span>        links {
</span><span>            Net(</span><span style="color:#f29668;">::</span><span>net</span><span style="color:#f29668;">::</span><span>Error</span><span style="color:#bfbab0cc;">, </span><span style="color:#f29668;">::</span><span>net</span><span style="color:#f29668;">::</span><span>ErrorKind)</span><span style="color:#bfbab0cc;">;
</span><span>        }
</span><span>    }
</span><span>
</span><span>    </span><span style="color:#ff7733;">pub fn </span><span style="color:#ffb454;">run</span><span>() </span><span style="color:#bfbab0cc;">-&gt; </span><span style="font-style:italic;color:#39bae6;">Result</span><span>&lt;()&gt; {
</span><span>        </span><span style="color:#ff7733;">match </span><span style="color:#f07178;">do_something</span><span>() {
</span><span>            </span><span style="font-style:italic;color:#39bae6;">Err</span><span>(ErrorKind</span><span style="color:#f29668;">::</span><span>Net(</span><span style="color:#f29668;">::</span><span>net</span><span style="color:#f29668;">::</span><span>ErrorKind</span><span style="color:#f29668;">::</span><span>Utils(</span><span style="color:#f29668;">::</span><span>utils</span><span style="color:#f29668;">::</span><span>ErrorKind</span><span style="color:#f29668;">::</span><span>Parse)</span><span style="color:#bfbab0cc;">, </span><span style="color:#f29668;">_</span><span>)) </span><span style="color:#f29668;">=&gt; </span><span>{
</span><span>                </span><span style="color:#f29668;">...                                                                                   
</span><span>            }
</span><span>            </span><span style="color:#f29668;">_ =&gt; </span><span>{ </span><span style="color:#f29668;">... </span><span>}
</span><span>        }
</span><span>    }
</span><span>}
</span></code></pre>
<p>Такая глубокая диспетчеризация ошибок случается нечасто, но когда это случается, хорошо, что цепочка ошибок упрощает задачу. Опять же, цепочка ошибок позволяет начать с простого, но дает вам возможности по мере роста вашего проекта.</p>
<h2 id="tl-dr">TL;DR</h2>
<p>Когда вы начинаете писать новое приложение на Rust, первое, что вы должны спросить, это «как я собираюсь обрабатывать ошибки?»; и ответ, вероятно, должен быть таким: «Я просто собираюсь настроить цепочку ошибок». Настройте цепочку ошибок, используя пример quickstart.rs, и ни о чем не беспокойтесь.</p>
<p>Дополнительные сведения о цепочке ошибок см. В документации. </p>










<script src="https://giscus.app/client.js"
    data-repo="dudochkin-victor&#x2F;dudochkin-victor.github.io"
    data-repo-id="MDEwOlJlcG9zaXRvcnkzNzgwNTYx"
    data-category="General"
    data-category-id="DIC_kwDOADmv0c4CAhvj"
    data-mapping="pathname"
    data-reactions-enabled="1"
    data-emit-metadata="0"
    data-theme="light"
    data-lang="en"
    crossorigin="anonymous"
    async>
</script>



        </div>

        
        
    </main>

    
    <footer>
        <small class="subtext">
            <a href="https://angular-rust.github.io">Dudochkin Victor</a> © 2021
        </small>
    </footer>
    
</body>
<script>
    function highlightNav(heading) {
        let pathname = location.pathname;
        document.querySelectorAll(".toc a").forEach((item) => {
            item.classList.remove("active");
        });
        document.querySelector(".toc a[href$='" + pathname + "#" + heading + "']").classList.add("active");
    }

    let currentHeading = "";
    window.onscroll = function () {
        let h = document.querySelectorAll("h1,h2,h3,h4,h5,h6");
        let elementArr = [];

        h.forEach(item => {
            if (item.id !== "") {
                elementArr[item.id] = item.getBoundingClientRect().top;
            }
        });
        elementArr.sort();
        for (let key in elementArr) {
            if (!elementArr.hasOwnProperty(key)) {
                continue;
            }
            if (elementArr[key] > 0 && elementArr[key] < 300) {
                if (currentHeading !== key) {
                    highlightNav(key);
                    currentHeading = key;
                }
                break;
            }
        }
    }
</script>

</html>
